Naf obj pre crop

namirfa opened the pull request:
Closes Issue #151 

Using the VANTIQ Source configuration, a user can "pre crop" all images before they are processed. "preCrop" values can be specified in query parameters as well, which will override those specified in the source configuration. Otherwise, the source config values will be used.

Edit: Updated the README to describe this new feature.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -227,6 +241,29 @@ private void setup(Map<String, ?> neuralNet, String sourceName, String modelDire
        }
+       
+       // Checking if pre cropping was specified in config
+       if (neuralNet.get("preCrop") instanceof Map) {
```
change name to `cropBeforeAnalysis` or something like that.

Variable name is fine, but like to have better name for configuration.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -291,6 +333,38 @@ public NeuralNetResults processImage(byte[] image, Map<String, ?> request) throw
+        boolean queryCrop = false;
+        int x, y, w, h;
+        x = y = w = h = -1;
+        if (request.get("preCrop") instanceof Map) {
```
ditto to name change for parameter

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -315,6 +389,45 @@ public NeuralNetResults processImage(byte[] image, Map<String, ?> request) throw
+     * @param h         The height of the "sub image"
+     * @return          The cropped image resulting from the parameters above
+     */
+    public byte[] cropImage(byte[] image, int x, int y, int w, int h) {
```
Cropping doesn't seem like something that's YOLO specific, is it?  Would this be better in some possibly abstract superclass that YOLO & any other processor could extend to get this capability?

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -315,6 +389,45 @@ public NeuralNetResults processImage(byte[] image, Map<String, ?> request) throw
+            // Convert byte[] to BufferedImage and crop
+            ByteArrayInputStream bais = new ByteArrayInputStream(image);
+            BufferedImage buffImage = ImageIO.read(bais);
+            BufferedImage croppedImage = buffImage.getSubimage(x, y, w, h);
```
do we know why we choose this mechanism (`BufferedImage`-based) instead of `OpenCV`-based?

I don't know which is right (if either), just wondering why this choice -- just because other things in YOLO are `AWT`-based?  Probably the correct choice, just wondering.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -315,6 +389,45 @@ public NeuralNetResults processImage(byte[] image, Map<String, ?> request) throw
+            
+            // Close input/output streams
+            baos.close();
+            bais.close();
```
Seems like either this should be in a `finally` block or done as a `try-with-resources`.  Otherwise, errors risk leaking open streams.

`try-with-resources` is fine with multiple resources...

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java:
```diff
@@ -69,6 +69,12 @@
+    static final int TOP_LEFT_X_COORDINATE = 50;
+    static final int TOP_LEFT_Y_COORDINATE = 50;
+    static final int CROPPED_WIDTH = 200;
+    static final int CROPPED_HEIGHT = 150;
```
Might want to name these constants with some PRECROP initial naming so it's obvious where they are used.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -291,6 +333,38 @@ public NeuralNetResults processImage(byte[] image, Map<String, ?> request) throw
+            if (preCrop.get("h") instanceof Integer && (Integer) preCrop.get("h") >= 0) {
+                h = (Integer) preCrop.get("h");
+            }
+            if (x >= 0 && y >= 0 && w >= 0 && h >= 0) {
```
I don't think we'll work in [Flatland](https://en.wikipedia.org/wiki/Flatland), so I think width & height need to be > 1, don't they?

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java:
```diff
@@ -1217,7 +1222,432 @@ public void testQuery() throws ImageProcessingException, InterruptedException {
+        // The coordinates are negative, should not do any resizing
+        preCrop.put("x", -1);
+        preCrop.put("y", -1);
+        preCrop.put("w", CROPPED_WIDTH);
```
may be the same code, but good to check these individually?  Also, check for 0 h & w, as well as missing (only, x, x & y, x&y&h, etc.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -227,6 +241,29 @@ private void setup(Map<String, ?> neuralNet, String sourceName, String modelDire
+           if (preCrop.get("y") instanceof Integer && (Integer) preCrop.get("y") >= 0) {
+               y = (Integer) preCrop.get("y");
+           }
+           if (preCrop.get("w") instanceof Integer && (Integer) preCrop.get("w") >= 0) {
```
`w` vs. `width`?  Or do we already have other things that are `w` & `h`?

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java:
```diff
@@ -1217,7 +1222,432 @@ public void testQuery() throws ImageProcessingException, InterruptedException {
+    }
+    
+    @Test
+    public void testPreCroppingLocal() throws ImageProcessingException, IOException {
```
Seems like we're testing the "crop" part pretty thoroughly, but not the "pre" part.

It would be good to ensure that the image is actually cropped before analysis.  May have to expose the TF results to the test methods so that you can verify that either it doesn't find anything (since everything interesting was cropped out) or that what it does find is within the boundaries of the cropped image.

If the `crop()` method is exposed to the test, you may be able to reach in and find the last image dimensions or something?

Given the structure, not completely sure how to do this, but seems like it'd be good to verify that the analysis is running on something cropped -- not just that the parameter validation is OK and that the resulting image matches those...

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -167,6 +167,12 @@ The Configuration document may look similar to the following example:
              "savedResolution": {
                 "longEdge": 400
+             },
+             "preCrop": {
```
Prefer to have something more descriptive like `cropBeforeAnalysis`.  (re-mentioned in the code which I started before the readme.

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -167,6 +167,12 @@ The Configuration document may look similar to the following example:
+             "preCrop": {
+                "x": 50,
+                "y": 100,
+                "w": 600,
```
`w` & `h` --> `width` & `height`?  unless we have other places where we use `w` & `h`

namirfa commented:
@fhcarter 

Added fixes for your reviewer comments.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/ImageCropper.java:
```diff
@@ -0,0 +1,52 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ImageCropper {
```
`ImageCropper` --> `NeuralNetBase` or something.  Other utils, etc., may get added later.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/YoloProcessor.java:
```diff
@@ -227,6 +241,29 @@ private void setup(Map<String, ?> neuralNet, String sourceName, String modelDire
+       
+       // Checking if pre cropping was specified in config
+       if (neuralNet.get("cropBeforeAnalysis") instanceof Map) {
+           Map preCrop = (Map) neuralNet.get("cropBeforeAnalysis");
```
String constants.

If you typed it > 1 time, it ought to be a constant.

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java:
```diff
@@ -69,6 +69,18 @@
+    static final int KEYBOARD_TOP_LEFT_X_COORDINATE = 120;
+    static final int KEYBOARD_TOP_LEFT_Y_COORDINATE = 250;
+    static final int KEYBOARD_CROPPED_WIDTH = 230;
+    static final int KEYBOARD_CROPPED_HEIGHT = 125;
```
same comment as above :-(

fhcarter replied:
Intent here was that these constants are used in the precrop tests, so shouldn't they have the same PRECROP prefix as those on line 72ish.

namirfa commented:
@fhcarter 

I renamed the `ImageCropper` class to `NeuralNetUtils`, and I added a constant for the `cropBeforeAnalysis` config option. I wasn't sure what you were referring to with your comment under `TestYoloProcessor`. 

I was planning on adding an issue to change all of the config option strings to constants, in both the `YoloProcessor` class and all of the tests as well. I figured this could be done as a separate PR, but I can also get it done now. Let me know what you think is better.
